pkg/features/features.go

```diff
@@ -0,0 +1,145 @@
+package features
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+// FeatureGate is feature gate 'manager' to be used
+var FeatureGate *featureGate
+
+type maturity string
+
+const (
+	// Alpha - alpha version
+	Alpha = maturity("ALPHA")
+	// Beta - beta version
+	Beta = maturity("BETA")
+	// GA - general availability
+	GA = maturity("GA")
+
+	// Deprecated - feature that will be deprecated in 2 releases
+	Deprecated = maturity("DEPRECATED")
+)
+
+// Feature name
+type feature string
+
+// List of supported features
+const (
+	// AlphaFeature - description
+	AlphaFeature feature = "alphaFeature"
+
+	// BetaFeature - description
+	BetaFeature feature = "betaFeature"
+
+	// GaFeature - description
+	GaFeature feature = "gaFeature"
+
+	// DeprecatedFeature - description
+	DeprecatedFeature feature = "deprecatedFeature"
+)
+
+type featureSpec struct {
+	// Default is the default enablement state for the feature
+	Default bool
+	// Maturity indicates the maturity level of the feature
+	Maturity maturity
+}
+
+var defaultSriovDpFeatureGates = map[feature]featureSpec{
+	AlphaFeature:      {Default: false, Maturity: Alpha},
+	BetaFeature:       {Default: true, Maturity: Beta},
+	GaFeature:         {Default: true, Maturity: GA},
+	DeprecatedFeature: {Default: false, Maturity: Deprecated},
+}
+
+type featureGate struct {
+	knownFeatures map[feature]featureSpec
+	enabled       map[feature]bool
+}
+
+func init() {
+	FeatureGate = newFeatureGate()
+}
+
+func newFeatureGate() *featureGate {
+	if FeatureGate != nil {
+		return FeatureGate
+	}
+	fg := &featureGate{}
+	fg.knownFeatures = make(map[feature]featureSpec)
+	fg.enabled = make(map[feature]bool)
+
+	for k, v := range defaultSriovDpFeatureGates {
+		fg.knownFeatures[k] = v
+	}
+
+	for k, v := range fg.knownFeatures {
+		fg.enabled[k] = v.Default
+	}
+	return fg
+}
+
+// Enabled returns enabelement status of the provided feature
+func (fg *featureGate) Enabled(f feature) bool {
+	return FeatureGate.enabled[f]
+}
+
+func (fg *featureGate) isFeatureSupported(f feature) bool {
+	_, exists := fg.knownFeatures[f]
+	return exists
+}
+
+func (fg *featureGate) set(f feature, status bool) error {
+	if !fg.isFeatureSupported(f) {
+		return fmt.Errorf("Feature %s is not supported", f)
+	}
+	fg.enabled[f] = status
+	if status == true && fg.knownFeatures[f].Maturity == Deprecated {
```

Comment:
redundant comparison to true. Also, check if key in knownFeatures before indexing map.

Reply:
Did you mean that is thousl be just `if status &&'?

As for the key checking, it's done in `isFeatureSupported()` which is called earlier in the code.